### ISSUE TITLE
Istio 1.6: networking issue during pod initialziation

### ISSUE THREAD

**Issue by illidan80** [istio/istio#25493]
As part of my micro-service bootstrap process, I calculating whether dependent micro-services, which running on the same Kubernetes, are in "ready" state.

We were using this mechanics for last year and all this time with Istio / auto injected sidecars (starting from version 1.1 and so on). It always worked well.

Starting with version 1.6 we experiencing issues with accessing Kubernetes API during startup. It looks like something changed in networking and my guess it related to initialization which takes now longer. Trying to access Kubernetes API during pod initialization may produce the following error:

```
io.kubernetes.client.openapi.ApiException: java.net.ConnectException: Failed to connect to /10.152.183.1:443
	at io.kubernetes.client.openapi.ApiClient.execute(ApiClient.java:898)
	at io.kubernetes.client.openapi.apis.CoreV1Api.listNamespacedPodWithHttpInfo(CoreV1Api.java:17358)
	at io.kubernetes.client.openapi.apis.CoreV1Api.listNamespacedPod(CoreV1Api.java:17329)
```

The issue is not permanent, looks after several retries eventually it succeed to call to Kubernetes API

UPDATE: I noticed that I can't connect database as well. Looks like during initialization there is a global issue related to networking

I using Kubernetes 1.17
Istio 1.6.4
Microk8s

**Comment by GregHanson** [comment]
There is usually some sort of delay from pod creation to when the app pod can successfully make outbound calls.  The init container sets up iptables rules to route all traffic to the istio sidecar proxy container and there is a delay caused by envoy waiting to start and receive configs from Istio control plane.  So retries / delays / exponential fallback are recommended for startup connections for apps.

What startup times are you seeing in 1.6 and how do those compare to the startup times you were seeing prior to this?

**Comment by illidan80** [comment]
@GregHanson I don't have the exact numbers. Before - as soon as I did "apply" to the deployment - everything was up successfully. Now when I deploying I have multiple errors in many services, related to failed network requests during initialization.

Retries is problematic since there are many places where such retries will be needed. The calls are not only to the kubernets, but also to external services (databases, etc).

Is there a deterministic way to know, from within container, that initialization finished?

**Comment by GregHanson** [comment]
@illidan80 there is a new feature that will be available in 1.7 that might be worth checking out:
https://github.com/istio/istio/pull/24737

**Comment by howardjohn** [comment]
Closing as resolved/mitigated by https://github.com/istio/istio/pull/24737. At the very least its a known bug actively tracked in https://github.com/istio/istio/issues/11130
